gg_fit_curve.py

```diff
@@ -0,0 +1,402 @@
+    x: float
+    y: float
+
+    def magnitude(self) -> float:
+        return sqrt(self.x * self.x + self.y * self.y)
+
+    # V2AddII; it's weird this works on vectors...
+    def add(self, other) -> "Point":
+        return Point(self.x + other.x, self.y + other.y)
+
+    # V2SubII
+    def sub(self, other) -> "Point":
+        return Point(self.x - other.x, self.y - other.y)
+
+    def scale(self, s: float) -> "Point":
+        return Point(self.x * s, self.y * s)
+
+    @staticmethod
+    def tan_left(points: Sequence["Point"]) -> "Vector":
+        assert len(points) > 1
+        return Vector.p0_to_p1(points[0], points[1]).normalize()
+
+    @staticmethod
+    def tan_right(points: Sequence["Point"]) -> "Vector":
+        assert len(points) > 1
+        return Vector.p0_to_p1(points[-1], points[-2]).normalize()
+
+    @staticmethod
+    def tan_center(points: Sequence["Point"]) -> "Vector":
+        assert len(points) == 3
+        v1 = Vector.p0_to_p1(points[1], points[0])
+        v2 = Vector.p0_to_p1(points[2], points[1])
+        tHatCenter = Vector((v1.x + v2.x) / 2, (v1.y + v2.y) / 2)
+        return tHatCenter.normalize()
+
+
+CubicBezier = Tuple[Point, Point, Point, Point]
+
+
+class Vector(Point):
+    def squared_length(self) -> float:
+        return self.x * self.x + self.y * self.y
+
+    def length(self) -> float:
+        return sqrt(self.squared_length())
+
+    def normalize(self) -> "Vector":
+        length = self.length()
+        x, y = 0.0, 0.0
+        if length != 0.0:
+            x = self.x / length
+            y = self.y / length
+
+        return Vector(x, y)
+
+    def negate(self) -> "Vector":
+        return Vector(-self.x, -self.y)
+
+    @staticmethod
+    def p0_to_p1(p0: Point, p1: Point) -> "Vector":
+        return Vector(*(p1.sub(p0)))
+
+    @staticmethod
+    def dot_product(v0: "Vector", v1: "Vector") -> float:
+        return v0.x * v1.x + v0.y * v1.y
+
+
+def degree(curve):
+    deg = len(curve) - 1
+    assert deg in range(1, 4), f"we don't expect to handle degrees outside [1,3]: {deg}"
+    return deg
+
+
+def derivative(curve):
+    curve_degree = degree(curve)
+    deriv = []
+    for i in range(curve_degree):
+        deriv.append(
+            Point(
+                (curve[i + 1].x - curve[i].x) * curve_degree,
+                (curve[i + 1].y - curve[i].y) * curve_degree,
+            )
+        )
+    return tuple(deriv)
+
+
+#  Reparameterize:
+#     Given set of points and their parameterization, try to find
+#     a better parameterization.
+def _reparameterize(points, first, last, u, curve):
+    new_ts = []
+    for i in range(first, last + 1):
+        new_ts.append(newton_raphson_root(curve, points[i], u[i - first]))
+    return tuple(new_ts)
+
+
+def newton_raphson_root(curve, pt, u):
+    # Compute Q(u)
+    curve_pt = point_at_t(curve, u)
+
+    # Generate control vertices for curve'
+    # TODO == cubic derivative; should just have derivative(curve)?
```

Comment:
i don't understand this TODO. Yes, it's the first derivative of a cubic, because we know `curve` has 4 ctrl points hence it's a cubic

Reply:
I think the TODO is actually done and that initially the code actually computed the derivative instead of calling derivative()